File: app/src/main/java/com/martinlaizg/geofind/views/fragment/creator/CreatePlaceFragment.java
package com.martinlaizg.geofind.views.fragment.creator;

import android.app.AlertDialog;
import android.content.Context;
import android.location.Address;
import android.location.Geocoder;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.OnBackPressedCallback;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.Navigation;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.textfield.TextInputLayout;
import com.martinlaizg.geofind.R;
import com.martinlaizg.geofind.data.access.database.entities.Place;
import com.martinlaizg.geofind.views.viewmodel.CreatorViewModel;
import com.squareup.picasso.Picasso;

import java.io.IOException;
import java.util.List;
import java.util.Objects;

import butterknife.BindView;
import butterknife.ButterKnife;

public class CreatePlaceFragment
		extends Fragment
		implements View.OnClickListener, OnMapReadyCallback {

	public static final String PLACE_POSITION = "PLACE_POSITION";

	private static final String TAG = CreatePlaceFragment.class.getSimpleName();

	private static final int CAMERA_UPDATE_ZOOM = 15;

	@BindView(R.id.new_place_name_layout)
	TextInputLayout new_place_name;
	@BindView(R.id.new_place_description_layout)
	TextInputLayout new_place_description;
	@BindView(R.id.new_place_create_button)
	MaterialButton create_button;
	@BindView(R.id.new_place_map_view)
	MapView new_place_map_view;
	@BindView(R.id.new_place_address)
	TextView new_place_address;
	@BindView(R.id.new_place_location_button)
	MaterialButton new_place_location_button;

	@BindView(R.id.new_place_image_button)
	MaterialButton new_place_image_button;
	@BindView(R.id.place_image_view)
	ImageView place_image_view;

	@BindView(R.id.question_switch)
	Switch question_switch;
	@BindView(R.id.question_layout)
	ConstraintLayout question_layout;
	@BindView(R.id.new_place_question)
	TextInputLayout new_place_question;
	@BindView(R.id.new_place_correct_answer)
	TextInputLayout new_place_correct_answer;
	@BindView(R.id.new_place_answer_2)
	TextInputLayout new_place_answer_2;
	@BindView(R.id.new_place_answer_3)
	TextInputLayout new_place_answer_3;

	private CreatorViewModel viewModel;
	private GoogleMap googleMap;
	private String image_url;

	@Override
	public void onClick(View v) {
		// Hide the keyboard
		InputMethodManager editTextInput = (InputMethodManager) requireActivity()
				.getSystemService(Context.INPUT_METHOD_SERVICE);
		View currentFocus = requireActivity().getCurrentFocus();
		if(currentFocus != null) {
			editTextInput.hideSoftInputFromWindow(currentFocus.getWindowToken(), 0);
		}

		if(!storePlace()) return;

		viewModel.savePlace();
		Navigation.findNavController(requireActivity(), R.id.main_fragment_holder).popBackStack();
	}

	/**
	 * Sotre the place data (name, description, image, question) into the viewModel {@link Place}
	 *
	 * @return true if no has errors else false
	 */
	private boolean storePlace() {
		// Get the name
		String placeName = Objects.requireNonNull(new_place_name.getEditText()).getText().toString()
				.trim();
		if(placeName.isEmpty()) {
			new_place_name.setError(getString(R.string.required_name));
			return false;
		}
		if(placeName.length() > getResources().getInteger(R.integer.max_name_length)) {
			new_place_name.setError(getString(R.string.text_too_long));
			return false;
		}
		new_place_name.setError("");

		// Get the description
		String placeDescription = Objects.requireNonNull(new_place_description.getEditText())
				.getText().toString().trim();
		if(placeDescription.isEmpty()) {
			new_place_description.setError(getString(R.string.required_description));
			return false;
		}
		if(placeDescription.length() >
				getResources().getInteger(R.integer.max_description_length)) {
			new_place_description.setError(getString(R.string.text_too_long));
			return false;
		}
		new_place_description.setError("");

		// Get the question
		boolean question = storeQuestion();
		if(!question) return false;

		viewModel.getPlace().setName(placeName);
		viewModel.getPlace().setDescription(placeDescription);

		return true;
	}

	/**
	 * Store the question data into the viewModel {@link Place}
	 *
	 * @return true if no has errors else false
	 */
	private boolean storeQuestion() {
		if(!question_switch.isChecked()) {
			viewModel.getPlace().setQuestion(null);
			viewModel.getPlace().setAnswer(null);
			viewModel.getPlace().setAnswer2(null);
			viewModel.getPlace().setAnswer3(null);
			return true;
		}

		// Check the question
		String question = Objects.requireNonNull(new_place_question.getEditText()).getText()
				.toString().trim();
		new_place_question.setError("");
		if(question.isEmpty()) {
			new_place_question.setError(getString(R.string.required_question));
			return false;
		}

		// Check the correct answer
		String correctAnswer = Objects.requireNonNull(new_place_correct_answer.getEditText())
				.getText().toString().trim();
		new_place_correct_answer.setError("");
		if(correctAnswer.isEmpty()) {
			new_place_correct_answer.setError(getString(R.string.required_answer));
			return false;
		}

		// Check the second answer
		String secondAnswer = Objects.requireNonNull(new_place_answer_2.getEditText()).getText()
				.toString().trim();
		new_place_answer_2.setError("");
		if(secondAnswer.isEmpty()) {
			new_place_answer_2.setError(getString(R.string.required_answer));
			return false;
		}

		// Check the third answer
		String thirdAnswer = Objects.requireNonNull(new_place_answer_3.getEditText()).getText()
				.toString().trim();
		new_place_answer_3.setError("");
		if(thirdAnswer.isEmpty()) {
			new_place_answer_3.setError(getString(R.string.required_answer));
			return false;
		}

		if(secondAnswer.equals(thirdAnswer)) {
			new_place_answer_3.setError(getString(R.string.answer_repeated));
			return false;
		}

		// Set the values
		viewModel.getPlace()
				.setCompleteQuestion(question, correctAnswer, secondAnswer, thirdAnswer);

		return true;
	}

	@Override
	public void onMapReady(GoogleMap googleMap) {
		this.googleMap = googleMap;
		this.googleMap.getUiSettings().setAllGesturesEnabled(false);
		this.googleMap.clear();
		setPosition(viewModel.getPlace());
	}

	private void setPosition(Place place) {
		if(place == null) return;
		LatLng position = place.getPosition();
		if(position == null) return;

		// Change button text
		new_place_location_button.setText(getString(R.string.update));
		new_place_address.setVisibility(View.VISIBLE);

		// Get address
		Geocoder gc = new Geocoder(requireContext());
		List<Address> locations;
		new_place_address
				.setText(getString(R.string.two_csv, position.latitude, position.longitude));
		try {
			locations = gc.getFromLocation(position.latitude, position.longitude, 1);
			if(locations != null && locations.size() > 1) {
				new_place_address.setText(locations.get(0).getAddressLine(0));
			}
		} catch(IOException e) {
			e.printStackTrace();
		}

		if(googleMap != null) {
			// Set static map
			MarkerOptions marker = new MarkerOptions().position(position);
			googleMap.moveCamera(CameraUpdateFactory.newLatLngZoom(position, CAMERA_UPDATE_ZOOM));
			googleMap.addMarker(marker);
		}
	}

	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
			Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_create_place, container, false);
		ButterKnife.bind(this, view);
		// Back button listener
		requireActivity().getOnBackPressedDispatcher()
				.addCallback(this, new OnBackPressedCallback(true) {
					@Override
					public void handleOnBackPressed() {
						showExitDialog();
					}
				});
		return view;
	}

	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
		// Show/hide question layout
		question_switch.setOnCheckedChangeListener((buttonView, isChecked) -> question_layout
				.setVisibility(isChecked ?
						               View.VISIBLE :
						               View.GONE));

		viewModel = ViewModelProviders.of(requireActivity()).get(CreatorViewModel.class);
		Bundle b = getArguments();
		if(b != null) {
			int position = b.getInt(PLACE_POSITION, viewModel.getPlaces().size());
			viewModel.retrievePlace(position);
			setPlace();
		}
		create_button.setOnClickListener(this);
		new_place_image_button.setOnClickListener(v -> {
			AlertDialog alertDialog = buildDialog();
			alertDialog.show();
		});

		new_place_location_button.setOnClickListener(v -> {
			if(!storePlace()) return;

			// Hide the keyboard
			InputMethodManager editTextInput = (InputMethodManager) requireActivity()
					.getSystemService(Context.INPUT_METHOD_SERVICE);
			View currentFocus = requireActivity().getCurrentFocus();
			if(currentFocus != null) {
				editTextInput.hideSoftInputFromWindow(currentFocus.getWindowToken(), 0);
			}

			Navigation.findNavController(requireActivity(), R.id.main_fragment_holder)
					.navigate(R.id.toCreatePlaceLocation);
		});
	}

	private void setPlace() {
		Place place = viewModel.getPlace();
		if(place == null) {
			Toast.makeText(requireContext(), getString(R.string.something_went_wrong),
			               Toast.LENGTH_SHORT).show();
			Navigation.findNavController(requireActivity(), R.id.main_fragment_holder)
					.popBackStack();
			return;
		}

		// Set name
		if(place.getName() != null) {
			Objects.requireNonNull(new_place_name.getEditText()).setText(place.getName());
		}

		// Set description
		if(place.getDescription() != null) {
			Objects.requireNonNull(new_place_description.getEditText())
					.setText(place.getDescription());
		}
		// Set location marker
		setPosition(place);

		// Set question
		question_switch.setChecked(false);
		if(place.getQuestion() != null) {
			question_switch.setChecked(true);
			Objects.requireNonNull(new_place_question.getEditText()).setText(place.getQuestion());
			Objects.requireNonNull(new_place_correct_answer.getEditText())
					.setText(place.getAnswer());
			Objects.requireNonNull(new_place_answer_2.getEditText()).setText(place.getAnswer2());
			Objects.requireNonNull(new_place_answer_3.getEditText()).setText(place.getAnswer3());
		}

		// Set image
		place_image_view.setVisibility(View.GONE);
		if(place.getImage() != null && !place.getImage().isEmpty()) {
			Picasso.with(requireContext()).load(image_url).into(place_image_view);
			place_image_view.setVisibility(View.VISIBLE);
		}
	}

	private AlertDialog buildDialog() {
		AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
		// Get the layout inflater
		LayoutInflater inflater = requireActivity().getLayoutInflater();

		View view = inflater.inflate(R.layout.add_image_layout, new LinearLayout(requireContext()));
		TextInputLayout image_url_layout = view.findViewById(R.id.image_url_layout);
		Objects.requireNonNull(image_url_layout.getEditText()).setText(image_url);
		return builder.setView(view).setPositiveButton(R.string.ok, (dialog, id) -> {
			image_url = Objects.requireNonNull(image_url_layout.getEditText()).getText().toString();
			place_image_view.setVisibility(View.GONE);
			if(!image_url.isEmpty()) {
				Picasso.with(requireContext()).load(image_url).into(place_image_view);
				place_image_view.setVisibility(View.VISIBLE);
			}

			// Hide the keyboard
			InputMethodManager editTextInput = (InputMethodManager) requireActivity()
					.getSystemService(Context.INPUT_METHOD_SERVICE);
			View currentFocus = requireActivity().getCurrentFocus();
			if(currentFocus != null) {
				editTextInput.hideSoftInputFromWindow(currentFocus.getWindowToken(), 0);
			}
		}).create();
	}

	private void showExitDialog() {
		new MaterialAlertDialogBuilder(requireContext()).setTitle(R.string.are_you_sure)
				.setMessage(getString(R.string.exit_lose_data_alert))
				.setPositiveButton(getString(R.string.ok), (dialog, which) -> {
					Navigation.findNavController(requireActivity(), R.id.main_fragment_holder)
							.popBackStack();
				}).show();
	}
}

File: app/src/main/java/com/martinlaizg/geofind/views/fragment/creator/CreatePlaceFragment.kt
package com.martinlaizg.geofind.views.fragment.creator

import android.app.AlertDialog
import android.content.DialogInterface
import android.location.Address
import android.location.Geocoder
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.activity.OnBackPressedCallback
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders
import androidx.navigation.Navigation
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.material.button.MaterialButton
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.textfield.TextInputLayout
import com.martinlaizg.geofind.R
import com.martinlaizg.geofind.data.access.database.entities.Place
import com.martinlaizg.geofind.utils.KeyboardUtils
import com.martinlaizg.geofind.views.viewmodel.CreatorViewModel
import java.io.IOException

class CreatePlaceFragment : Fragment(), View.OnClickListener, OnMapReadyCallback {

	private var newPlaceMapView: MapView? = null
	private var newPlaceAddress: TextView? = null
	private var placeImageView: ImageView? = null

	private var newPlaceName: TextInputLayout? = null
	private var newPlaceDescription: TextInputLayout? = null

	private var createButton: MaterialButton? = null
	private var newPlaceLocationButton: MaterialButton? = null
	private var newPlaceImageButton: MaterialButton? = null
	private var questionSwitch: Switch? = null

	private var questionLayout: ConstraintLayout? = null
	private var newPlaceQuestion: TextInputLayout? = null
	private var newPlaceCorrectAnswer: TextInputLayout? = null
	private var newPlaceAnswer2: TextInputLayout? = null
	private var newPlaceAnswer3: TextInputLayout? = null

	private var viewModel: CreatorViewModel? = null
	private var googleMap: GoogleMap? = null
	private var imageUrl: String? = null

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
	                          savedInstanceState: Bundle?): View? {
		val view = inflater.inflate(R.layout.fragment_create_place, container, false)

		newPlaceMapView = view.findViewById(R.id.new_place_map_view)
		newPlaceAddress = view.findViewById(R.id.new_place_address)
		placeImageView = view.findViewById(R.id.place_image_view)

		newPlaceName = view.findViewById(R.id.new_place_name_layout)
		newPlaceDescription = view.findViewById(R.id.new_place_description_layout)

		createButton = view.findViewById(R.id.new_place_create_button)
		newPlaceLocationButton = view.findViewById(R.id.new_place_location_button)
		newPlaceImageButton = view.findViewById(R.id.new_place_image_button)
		questionSwitch = view.findViewById(R.id.question_switch)

		questionLayout = view.findViewById(R.id.question_layout)
		newPlaceQuestion = view.findViewById(R.id.new_place_question)
		newPlaceCorrectAnswer = view.findViewById(R.id.new_place_correct_answer)
		newPlaceAnswer2 = view.findViewById(R.id.new_place_answer_2)
		newPlaceAnswer3 = view.findViewById(R.id.new_place_answer_3)

		// Back button listener
		requireActivity().onBackPressedDispatcher
				.addCallback(viewLifecycleOwner, object : OnBackPressedCallback(true) {
					override fun handleOnBackPressed() {
						showExitDialog()
					}
				})
		return view
	}

	override fun onClick(v: View) {
		KeyboardUtils.hideKeyboard(requireActivity())
		if (!storePlace()) return
		viewModel!!.savePlace()
		Navigation.findNavController(requireActivity(), R.id.main_fragment_holder).popBackStack()
	}

	/**
	 * Store the place data (name, description, image, question) into the viewModel [Place]
	 *
	 * @return true if no has errors else false
	 */
	private fun storePlace(): Boolean {
		// Get the name
		val placeName = newPlaceName!!.editText!!.text.toString().trim()
		if (placeName.isEmpty()) {
			newPlaceName!!.error = getString(R.string.required_name)
			return false
		}
		if (placeName.length > resources.getInteger(R.integer.max_name_length)) {
			newPlaceName!!.error = getString(R.string.text_too_long)
			return false
		}
		newPlaceName!!.error = ""

		// Get the description
		val placeDescription = newPlaceDescription!!.editText!!.text.toString().trim()
		if (placeDescription.isEmpty()) {
			newPlaceDescription!!.error = getString(R.string.required_description)
			return false
		}
		if (placeDescription.length > resources.getInteger(R.integer.max_description_length)) {
			newPlaceDescription!!.error = getString(R.string.text_too_long)
			return false
		}
		newPlaceDescription!!.error = ""

		// Get the question
		val question = storeQuestion()
		if (!question) return false
		viewModel!!.place!!.name = placeName
		viewModel!!.place!!.description = placeDescription
		return true
	}

	/**
	 * Store the question data into the viewModel [Place]
	 *
	 * @return true if no has errors else false
	 */
	private fun storeQuestion(): Boolean {
		if (!questionSwitch!!.isChecked) {
			viewModel!!.place!!.question = null
			viewModel!!.place!!.answer = null
			viewModel!!.place!!.answer2 = null
			viewModel!!.place!!.answer3 = null
			return true
		}

		// Check the question
		val question = newPlaceQuestion!!.editText!!.text.toString().trim()
		newPlaceQuestion!!.error = ""
		if (question.isEmpty()) {
			newPlaceQuestion!!.error = getString(R.string.required_question)
			return false
		}

		// Check the correct answer
		val correctAnswer = newPlaceCorrectAnswer!!.editText!!.text.toString().trim()
		newPlaceCorrectAnswer!!.error = ""
		if (correctAnswer.isEmpty()) {
			newPlaceCorrectAnswer!!.error = getString(R.string.required_answer)
			return false
		}

		// Check the second answer
		val secondAnswer = newPlaceAnswer2!!.editText!!.text.toString().trim()
		newPlaceAnswer2!!.error = ""
		if (secondAnswer.isEmpty()) {
			newPlaceAnswer2!!.error = getString(R.string.required_answer)
			return false
		}

		// Check the third answer
		val thirdAnswer = newPlaceAnswer3!!.editText!!.text.toString().trim()
		newPlaceAnswer3!!.error = ""
		if (thirdAnswer.isEmpty()) {
			newPlaceAnswer3!!.error = getString(R.string.required_answer)
			return false
		}
		if (secondAnswer == thirdAnswer) {
			newPlaceAnswer3!!.error = getString(R.string.answer_repeated)
			return false
		}

		// Set the values
		viewModel!!.place!!.fillQuestion(question, correctAnswer, secondAnswer, thirdAnswer)
		return true
	}

	override fun onMapReady(googleMap: GoogleMap) {
		this.googleMap = googleMap
		this.googleMap!!.uiSettings.setAllGesturesEnabled(false)
		this.googleMap!!.clear()
		setPosition(viewModel!!.place)
	}

	private fun setPosition(place: Place?) {
		if (place == null) return
		val position = place.position ?: return

		// Change button text
		newPlaceLocationButton!!.text = getString(R.string.update)
		newPlaceAddress!!.visibility = View.VISIBLE

		// Get address
		val gc = Geocoder(requireContext())
		val locations: List<Address>?
		newPlaceAddress!!.text = getString(R.string.two_csv, position.latitude, position.longitude)
		try {
			locations = gc.getFromLocation(position.latitude, position.longitude, 1)
			if (locations != null && locations.isNotEmpty()) {
				newPlaceAddress!!.text = locations[0].getAddressLine(0)
			}
		} catch (e: IOException) {
			e.printStackTrace()
		}
		if (googleMap != null) {
			// Set static map
			val marker = MarkerOptions().position(position)
			googleMap!!.moveCamera(CameraUpdateFactory.newLatLngZoom(position, CAMERA_UPDATE_ZOOM.toFloat()))
			googleMap!!.addMarker(marker)
		}
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		// Show/hide question layout
		questionSwitch!!.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
			questionLayout!!.visibility = if (isChecked) View.VISIBLE else View.GONE
		}
		viewModel = ViewModelProviders.of(requireActivity()).get(CreatorViewModel::class.java)
		val b = arguments
		if (b != null) {
			val position = b.getInt(PLACE_POSITION, viewModel!!.places.size)
			viewModel!!.retrievePlace(position)
			setPlace()
		}
		createButton!!.setOnClickListener(this)
		newPlaceImageButton!!.setOnClickListener {
			val alertDialog = buildDialog()
			alertDialog.show()
		}
		newPlaceLocationButton!!.setOnClickListener {
			if (!storePlace()) return@setOnClickListener
			KeyboardUtils.hideKeyboard(requireActivity())
			Navigation.findNavController(requireActivity(), R.id.main_fragment_holder)
					.navigate(R.id.toCreatePlaceLocation)
		}
	}

	private fun setPlace() {
		val place = viewModel!!.place
		if (place == null) {
			Toast.makeText(requireContext(), getString(R.string.something_went_wrong),
					Toast.LENGTH_SHORT).show()
			Navigation.findNavController(requireActivity(), R.id.main_fragment_holder)
					.popBackStack()
			return
		}

		// Set name
		newPlaceName!!.editText!!.setText(place.name)

		// Set description
		newPlaceDescription!!.editText!!.setText(place.description)
		// Set location marker
		setPosition(place)

		// Set question
		questionSwitch!!.isChecked = false
		if (place.question != null) {
			questionSwitch!!.isChecked = true
			newPlaceQuestion!!.editText!!.setText(place.question)
			newPlaceCorrectAnswer!!.editText!!.setText(place.answer)
			newPlaceAnswer2!!.editText!!.setText(place.answer2)
			newPlaceAnswer3!!.editText!!.setText(place.answer3)
		}

		// Set image
		placeImageView!!.visibility = View.GONE
		if (place.image != null && place.image!!.isNotEmpty()) {
			imageUrl = place.image
			placeImageView!!.visibility = View.VISIBLE
			TODO("load image")
		}
	}

	private fun buildDialog(): AlertDialog {
		val builder = AlertDialog.Builder(activity)
		// Get the layout inflater
		val inflater = requireActivity().layoutInflater
		val view = inflater.inflate(R.layout.add_image_layout, LinearLayout(requireContext()))
		val imageUrlLayout: TextInputLayout = view.findViewById(R.id.image_url_layout)
		imageUrlLayout.editText!!.setText(imageUrl)
		return builder.setView(view).setPositiveButton(R.string.ok) { _: DialogInterface?, _: Int ->
			imageUrl = imageUrlLayout.editText!!.text.toString()
			placeImageView!!.visibility = View.GONE
			if (imageUrl!!.isNotEmpty()) {
				TODO("load image")
			}
			KeyboardUtils.hideKeyboard(requireActivity())
		}.create()
	}

	private fun showExitDialog() {
		MaterialAlertDialogBuilder(requireContext()).setTitle(R.string.are_you_sure)
				.setMessage(getString(R.string.exit_lose_data_alert))
				.setPositiveButton(getString(R.string.ok)) { _: DialogInterface?, _: Int ->
					Navigation
							.findNavController(requireActivity(), R.id.main_fragment_holder)
							.popBackStack()
				}.show()
	}

	companion object {
		val PLACE_POSITION = "PLACE_POSITION"
		val CAMERA_UPDATE_ZOOM = 15
	}
}